primitives/message-lane/src/target_chain.rs

```diff
@@ -0,0 +1,101 @@
+}
+
+/// Instant message dispatch API.
+///
+/// Used to process incoming messages in the message-delivery transaction.
+pub trait MessageDispatch<Payload, Fee> {
+	/// Creates instance of message dispatch, which can spent at most `weight` on
+	/// messages dispatch. Spending more than weight is not allowed.
+	fn with_allowed_weight(weight: Weight) -> Self;
+	/// Return weight that is left at this moment. Initially it is `weight` from
+	/// `with_allowed_weight` call, but it should be decreased after every successful
+	/// `dispatch` call.
+	fn weight_left(&self) -> Weight;
+	/// Estimate dispatch weight.
+	///
+	/// This function must: (1) be instant and (2) return correct, but upper bound
+	/// of dispatch weight.
+	fn dispatch_weight(&self, message: &Message<Payload, Fee>) -> Weight;
+	/// Called to process inbound message.
+	///
+	/// It is up to the implementers of this trait to determine whether the message
+	/// is invalid (i.e. improperly encoded, has too large weight, ...) or not. And,
+	/// if message is invalid, then it should be dropped immediately (by returning
+	/// `MessageResult::Processed`), or it'll block the lane forever.
+	///
+	/// If `MessageDispatch` ever returns `MessageResult::NotProcessed`, then it should
+	/// remember that it has queued this message and provide `QueuedMessageDispatch`
+	/// implementation that uses this information later.
+	///
+	/// Decision whether to return `MessageResult::NotProcessed` from this function
+	/// **MUST** be made instantly. We do not account weight, spent to make this
+	/// decision.
+	fn dispatch(&mut self, message: Message<Payload, Fee>) -> MessageResult<Payload, Fee>;
+}
+
+/// Queued message dispatch API.
+pub trait QueuedMessageDispatch<Payload> {
+	/// Creates instance of message dispatch, which can spent at most `weight` on
+	/// messages dispatch. Spending more than weight is not allowed.
+	fn with_allowed_weight(weight: Weight) -> Self;
+	/// Creates instance of message dispatch, which could spent any weight on
+	/// dispatching messages. "Any" here means that it **MUST** fit the block
+	/// weight AND that it should not always take 100% of block weight (or it'll
+	/// block all other transactions from including into block).
+	fn with_any_weight() -> Self;
+	/// Called to process queued messages.
+	///
+	/// Returns number of processed queued messages and weight 'spent' on that.
+	fn dispatch(&mut self) -> (MessageNonce, Weight);
+}
+
+/// Message dispatch payment. It is called as a part of message-delivery transaction. Delivery
+/// transaction submitter (message relay account) pays for the messages dispatch (probably
+/// in advance).
+pub trait MessageDispatchPayment<AccountId> {
+	/// Error type.
+	type Error: Debug + Into<&'static str>;
+
+	/// Write-off fee for dispatching messages of given cumulative weight.
+	fn pay_dispatch_fee(payer: &AccountId, weight: Weight) -> Result<(), Self::Error>;
```

Comment:
Same with the other trait, this abstraction is imho unnecessary and we should bundle `Currency` trait directly with the message lane.

Reply:
I don't think so. We may provide some default implementations that are based on Currency trait, but we should not force everyone to use Currency trait.